transfer and index dialects added

Hatsunespica opened the pull request:
Hi Mathieu,

This is my new PR! It only contains two dialect definitions. It's easier to review!

Thanks!
Yuyou

math-fehr reviewed (commented):
Thanks!
Forget about the pyright errors that you got in the review, I'm currently fixing that.
Otherwise, for the PR, the `index_dialect` looks good.
For the `transfer` dialect, can we remove `NegOp` and `IfOp`? We don't need `NegOp` as we can use `SubOp` in `Index`, and we don't need `IfOp` since we can use the `If` in `scf`!

Hatsunespica commented:
Hi!

You are right, NegOP might be useless so I removed it. 
Besides, I also renamed IfOp with SelectOp to keep name convension. I need this op in integer analysis ....

math-fehr commented:
Thanks a lot! I'll merge this now!
